### PR TITLE
Add owner permission check and credential check

### PR DESCRIPTION
**Pull request by jiang95-dev**
1. Add auth login when terraform init fails.
2. Add check on owner permission of a project.
3. Remove terraform warnings.
4. Adjust log wordings.

### CODE REVIEW

File: terraform/03_gke_cluster.tf [hunk 1]
```diff
@@ -67,8 +67,6 @@ resource "google_container_cluster" "gke" {
   # interesting things.
   node_pool {
     node_config {
-      machine_type = "n1-standard-2"
```
Comment (simonz130) [inline terraform/03_gke_cluster.tf:70]: Don't we need to specify the machine_type anymore?

Comment (simonz130) [reply terraform/03_gke_cluster.tf:70]: @Daniel-Sanche - was there any change related to this?

File: terraform/install.sh [hunk 2]
```diff
@@ -54,7 +54,7 @@ getBillingAccount() {
   done
 
   if [[ $(echo "${found_accounts}" | wc -l) -gt 1 ]]; then
-      log "Which billing account would you like to use?:"
+      log "Which billing account would you like to use? Enter the number next to the billing account:"
```
Comment (simonz130) [inline terraform/install.sh:57]: Enter the number representing a billing account you want to use

File: terraform/install.sh [hunk 3]
```diff
@@ -75,22 +75,35 @@ getBillingAccount() {
 }
 
 getProject() {
-  log "Checking for project list"
+  log "Checking for project list..."
+  acct=$(gcloud info --format="value(config.account)")
+  # get projects with user account
+  owner_projects=$(gcloud projects list --filter="id:stackdriver-sandbox-* AND name='Stackdriver Sandbox Demo'" --format="value(projectId)")
```
Comment (simonz130) [inline terraform/install.sh:81]: Where is the filter for owned projects? I see you filter the projects that have Stackdriver Sandbox Demo in their name

Comment (simonz130) [reply terraform/install.sh:81]: I see that you make the determination of owner later on. The name owner_projects here is then misleading.

File: terraform/install.sh [hunk 4]
```diff
@@ -75,22 +75,35 @@ getBillingAccount() {
...
+      fi
     fi
   done
-
+  # make user choose projects
   if [ -z "$found_projects" ] || [[ ${#found_projects[@]} -eq 0 ]]; then
     createProject;
   else
-      log "Which project would you like to use?:"
+      log "Which project would you like to use? Enter the number next to the project:"
```
Comment (simonz130) [inline terraform/install.sh:106]: Enter the number next to the project you would like to use:

File: terraform/install.sh [hunk 5]
```diff
@@ -236,7 +255,7 @@ log "Checking Prerequisites..."
 getBillingAccount;
 
 log "Install current version of Terraform"
-installTerraform
+#installTerraform
```
Comment (simonz130) [inline terraform/install.sh:258]: You can remove this given that now it's part of the Cloud Shell container image

File: terraform/03_gke_cluster.tf (hunk 1 above)
Comment (jiang95-dev) [reply terraform/03_gke_cluster.tf:70]: It didn't modify this line.

File: terraform/install.sh (hunk 5 above)
Comment (jiang95-dev) [reply terraform/install.sh:258]: Ok.

File: terraform/03_gke_cluster.tf (hunk 1 above)
Comment (Daniel-Sanche) [reply terraform/03_gke_cluster.tf:70]: We do need this line. Please make sure you keep changes from upstream

File: terraform/03_gke_cluster.tf [hunk 6]
```diff
@@ -78,8 +76,8 @@ resource "google_container_cluster" "gke" {
         cluster = "stackdriver-sandbox-main"   
       }
     }
-
-    initial_node_count = 4
+    
+    initial_node_count = 7
```
Comment (Daniel-Sanche) [inline terraform/03_gke_cluster.tf:80]: Revert this

File: terraform/destroy.sh [hunk 7]
```diff
@@ -26,7 +26,7 @@ SCRIPT_DIR=$(realpath $(dirname "$0"))
 cd $SCRIPT_DIR
 
 # find the stackdriver sandbox project id
-found=$(gcloud projects list --filter="name:stackdriver-sandbox-*" --format="value(projectId)")
+found=$(gcloud projects list --filter="id:stackdriver-sandbox-* AND name='Stackdriver Sandbox Demo'" --format="value(projectId)")
```
Comment (Daniel-Sanche) [inline terraform/destroy.sh:29]: revert this

Comment (jiang95-dev) [reply terraform/destroy.sh:29]: Oh, why is that? Shouldn't we also filter the name for demo?

File: terraform/00_state.tf [hunk 8]
```diff
@@ -26,6 +26,6 @@ terraform {
 data "terraform_remote_state" "state" {
     backend = "gcs"
     config = {
-        bucket = "${var.bucket_name}"
+        bucket = var.bucket_name
```
Comment (Daniel-Sanche) [inline terraform/00_state.tf:29]: Are all of these quote changes on purpose? What's the difference?

Comment (jiang95-dev) [reply terraform/00_state.tf:29]: To remove the warnings. Quotes are deprecated.

File: terraform/install.sh [hunk 9]
```diff
@@ -75,22 +75,35 @@ getBillingAccount() {
...
+  # get projects with user account
+  owner_projects=$(gcloud projects list --filter="id:stackdriver-sandbox-* AND name='Stackdriver Sandbox Demo'" --format="value(projectId)")
+  # get projects with billing account
   billed_projects=$(gcloud beta billing projects list --billing-account="$billing_id" --filter="project_id:stackdriver-sandbox-*" --format="value(projectId)")
-  # only keep projects with name "Stackdriver Sandbox Demo"
+  declare -A proj_map
+  for proj in ${billed_projects[@]}
+  do
+    proj_map["$proj"]="yes"
+  done
```
Comment (Daniel-Sanche) [inline terraform/install.sh:88]: I don't think I understand the purpose of proj_map

File: terraform/install.sh [hunk 10]
```diff
@@ -75,22 +75,35 @@ getBillingAccount() {
 }
 
 getProject() {
-  log "Checking for project list"
+  log "Checking for project list..."
+  acct=$(gcloud info --format="value(config.account)")
+  # get projects with user account
+  owner_projects=$(gcloud projects list --filter="id:stackdriver-sandbox-* AND name='Stackdriver Sandbox Demo'" --format="value(projectId)")
+  # get projects with billing account
   billed_projects=$(gcloud beta billing projects list --billing-account="$billing_id" --filter="project_id:stackdriver-sandbox-*" --format="value(projectId)")
```
Comment (Daniel-Sanche) [inline terraform/install.sh:83]: Why do you need "owner_projects" and "billed_projects"?

File: terraform/install.sh (hunk 9 above)
Comment (jiang95-dev) [reply terraform/install.sh:88]: It's easier to check if a variable exists in a map than a list.

File: terraform/install.sh (hunk 10 above)
Comment (jiang95-dev) [reply terraform/install.sh:83]: Because "gcloud projects list" won't tell you anything about the billing info. And not all billing project lets you get IAM if you don't have at least browser permission. So I have to combine them.

File: terraform/destroy.sh (hunk 7 above)
Comment (jiang95-dev) [reply terraform/destroy.sh:29]: @Daniel-Sanche

File: terraform/install.sh (hunk 10 above)
Comment (Daniel-Sanche) [reply terraform/install.sh:83]: Since we only care about projects with owner permission, do we need the ones without browser permission?

File: terraform/destroy.sh (hunk 7 above)
Comment (Daniel-Sanche) [reply terraform/destroy.sh:29]: sorry, I thought this was another accidental overwrite. This can be kept

File: terraform/00_state.tf (hunk 8 above)
Comment (Daniel-Sanche) [reply terraform/00_state.tf:29]: ok sounds good

File: terraform/install.sh (hunk 9 above)
Comment (Daniel-Sanche) [reply terraform/install.sh:88]: It would be more readable if we only had one list of projects, and we could just iterate through the list. Something like this: https://stackoverflow.com/questions/2696055/intersection-of-two-lists-in-bash

But I'm skeptical that we really need both `billed_projects` and `owned_projects`

File: terraform/install.sh (hunk 10 above)
Comment (jiang95-dev) [reply terraform/install.sh:83]: I mean if you use "gcloud projects get-iam-policy" on projects that you don't have at least browser permission, it will throw an error. And there are many projects under the billing account that you may not have permission for the current user account.

File: terraform/install.sh (hunk 5 above)
Comment (Daniel-Sanche) [reply terraform/install.sh:258]: @simonz130 aren't we leaving the cloud shell container out for this release though?

File: terraform/install.sh (hunk 10 above)
Comment (Daniel-Sanche) [reply terraform/install.sh:83]: We can ignore projects that throw an error though, right? We know those won't make it onto the list we display to the user

File: terraform/install.sh [hunk 11]
```diff
@@ -165,7 +165,8 @@ applyTerraform() {
   rm -f .terraform/terraform.tfstate
 
   log "Initialize terraform backend with bucket ${bucket_name}"  
-  if [[ $(terraform init -backend-config "bucket=${bucket_name}" -lock=false) ]]; then
+  
+  if [[ $(terraform init -backend-config "bucket=${bucket_name}" -lock=false 2> /dev/null | cat) ]]; then
```
Comment (Daniel-Sanche) [inline terraform/install.sh:169]: Won't this always return true now?

File: terraform/install.sh (hunk 5 above)
Comment (simonz130) [reply terraform/install.sh]: No, we are adding cloud shell custom image in upcoming release - the Cloud Shell team did fix the blocker on their side.

File: terraform/install.sh (hunk 11 above)
Comment (Daniel-Sanche) [reply terraform/install.sh:169]: `If [[ ` statements expect some kind of comparitor, this line as is may give us unexpected results

**Comment by jiang95-dev** [comment]
Reopen it after rebasing.

File: terraform/install.sh [hunk 12]
```diff
@@ -157,49 +161,24 @@ createProject() {
...
-  log "Please create a monitoring workspace for the project by clicking on the following link: $gcp_monitoring_path"
-  read -p "When you are done, please press enter to continue"
-
-  log "Creating monitoring examples (dashboards, uptime checks, alerting policies, etc.)..."
-  pushd monitoring/
-  terraform init -lock=false
-  terraform apply --auto-approve -var="project_id=${project_id}" -var="external_ip=${external_ip}"
-  popd
-}
-
```
Comment (simonz130) [inline terraform/install.sh:202]: why is this removed?
